File: examples/typescript/ai-agent-anthropic.ts
/**
 * AI Agent with Baldrick — using Anthropic's Claude to write and execute code.
 *
 * This example shows the "code as tool use" pattern:
 * 1. Claude writes TypeScript code that calls tools (external functions)
 * 2. Baldrick executes the code in a sandbox
 * 3. When the code calls a tool, Baldrick suspends and returns a snapshot
 * 4. Your app resolves the tool call, then resumes Baldrick with the result
 *
 * Prerequisites:
 *   npm install @anthropic-ai/sdk @baldrick/core
 *   export ANTHROPIC_API_KEY=sk-...
 *
 * Run with: npx ts-node examples/typescript/ai-agent-anthropic.ts
 */

import Anthropic from "@anthropic-ai/sdk";
import { Baldrick, BaldrickSnapshotHandle } from "@baldrick/core";

// --- Tool implementations (the real functions that run on your server) ---

async function getWeather(city: string): Promise<object> {
  // In production, call a real weather API
  const mockData: Record<string, object> = {
    London: { condition: "Partly cloudy", temp: 18, humidity: 72 },
    Tokyo: { condition: "Clear", temp: 26, humidity: 55 },
    "New York": { condition: "Rain", temp: 14, humidity: 88 },
  };
  return mockData[city] ?? { condition: "Unknown", temp: 0, humidity: 0 };
}

async function searchFlights(
  from: string,
  to: string,
  date: string
): Promise<object[]> {
  // In production, call a flight search API
  return [
    { airline: "BA", flight: "BA123", price: 450, departure: "08:00" },
    { airline: "AF", flight: "AF456", price: 380, departure: "14:30" },
  ];
}

// --- Map of available tools ---

const tools: Record<string, (...args: any[]) => Promise<any>> = {
  getWeather: (city: string) => getWeather(city),
  searchFlights: (from: string, to: string, date: string) =>
    searchFlights(from, to, date),
};

// --- The AI agent loop ---

async function runAgent(userQuery: string) {
  const client = new Anthropic();

  // Ask Claude to write code that answers the user's question
  const response = await client.messages.create({
    model: "claude-sonnet-4-20250514",
    max_tokens: 1024,
    system: `You are a helpful assistant. When the user asks a question that requires
external data (weather, flights, etc.), write TypeScript code that calls the
available functions and computes the answer.

Available functions (called with await):
- getWeather(city: string) → { condition: string, temp: number, humidity: number }
- searchFlights(from: string, to: string, date: string) → Array<{ airline: string, flight: string, price: number, departure: string }>

Available inputs: userQuery (the user's question as a string)

Return ONLY the TypeScript code, no markdown fences. The last expression is the output.`,
    messages: [{ role: "user", content: userQuery }],
  });

  const code =
    response.content[0].type === "text" ? response.content[0].text : "";
  console.log("Claude wrote:\n", code, "\n");

  // Execute the AI-generated code in Baldrick's sandbox
  const sandbox = new Baldrick(code, {
    inputs: ["userQuery"],
    externalFunctions: Object.keys(tools),
    timeLimitMs: 10000,
    memoryLimitMb: 16,
  });

  // Run with snapshot/resume loop to handle external function calls
  let state = sandbox.start({ userQuery });

  while (!state.completed) {
    const { functionName, args } = state;
    console.log(`  -> calling ${functionName}(${JSON.stringify(args)})`);

    // Look up the tool and call it with the arguments from the sandbox
    const toolFn = tools[functionName];
    if (!toolFn) {
      throw new Error(`Unknown tool: ${functionName}`);
    }
    const result = await toolFn(...args);
    console.log(`  <- ${functionName} returned:`, result);

    // Resume the sandbox with the tool's result
    const snapshot = BaldrickSnapshotHandle.load(state.snapshot);
    state = snapshot.resume(result);
  }

  console.log("\nFinal answer:", state.output);
  return state.output;
}

// --- Run it ---

async function main() {
  console.log("=== Example 1: Weather query ===\n");
  await runAgent("What's the weather like in Tokyo?");

  console.log("\n=== Example 2: Multi-tool query ===\n");
  await runAgent(
    "Compare the weather in London and New York, and find me cheap flights from London to New York for 2026-04-01."
  );
}

main().catch(console.error);

File: examples/python/ai-agent/ai_agent_anthropic.py
"""
AI Agent with Zapcode — LOW-LEVEL approach using Anthropic SDK directly.

This shows the manual snapshot/resume loop using the zapcode package.
For most use cases, prefer zapcode-ai instead — see ai_agent_zapcode_ai.py
for the recommended approach.

The "code as tool use" pattern:
1. Claude writes TypeScript code that calls tools (external functions)
2. Zapcode executes the code in a sandbox
3. When the code calls a tool, Zapcode suspends and returns a snapshot
4. Your app resolves the tool call, then resumes Zapcode with the result

Prerequisites:
    pip install anthropic zapcode
    # or: uv add anthropic zapcode

Run with: python ai_agent_anthropic.py
"""

import anthropic
from zapcode import Zapcode, ZapcodeSnapshot

# --- Tool implementations (the real functions that run on your server) ---

MOCK_WEATHER = {
    "London": {"condition": "Overcast", "temp": 12, "humidity": 80},
    "Tokyo": {"condition": "Clear", "temp": 26, "humidity": 55},
    "New York": {"condition": "Rain", "temp": 14, "humidity": 88},
}

MOCK_FLIGHTS = [
    {"airline": "BA", "flight": "BA123", "price": 450, "departure": "08:00"},
    {"airline": "AF", "flight": "AF456", "price": 380, "departure": "14:30"},
]


def get_weather(city: str) -> dict:
    """In production, call a real weather API."""
    return MOCK_WEATHER.get(city, {"condition": "Unknown", "temp": 0, "humidity": 0})


def search_flights(origin: str, destination: str, date: str) -> list[dict]:
    """In production, call a flight search API."""
    return MOCK_FLIGHTS


# Map of tool name -> implementation
TOOLS = {
    "getWeather": lambda city: get_weather(city),
    "searchFlights": lambda origin, dest, date: search_flights(origin, dest, date),
}

SYSTEM_PROMPT = """\
You are a helpful assistant. When the user asks a question that requires
external data (weather, flights, etc.), write TypeScript code that calls the
available functions and computes the answer.

Available functions (called with await):
- getWeather(city: string) → { condition: string, temp: number, humidity: number }
- searchFlights(origin: string, destination: string, date: string) → Array<{ airline: string, flight: string, price: number, departure: string }>

Available inputs: userQuery (the user's question as a string)

Return ONLY the TypeScript code, no markdown fences. The last expression is the output."""


def execute_in_sandbox(code: str, inputs: dict) -> any:
    """Execute AI-generated TypeScript in Zapcode's sandbox."""
    sandbox = Zapcode(
        code,
        inputs=list(inputs.keys()),
        external_functions=list(TOOLS.keys()),
        time_limit_ms=10_000,
    )

    state = sandbox.start(inputs)

    # Snapshot/resume loop: resolve each external call as it comes
    while state.get("suspended"):
        fn_name = state["function_name"]
        args = state["args"]
        print(f"  -> calling {fn_name}({args})")

        tool_fn = TOOLS.get(fn_name)
        if not tool_fn:
            raise ValueError(f"Unknown function: {fn_name}")

        result = tool_fn(*args)
        print(f"  <- {fn_name} returned: {result}")

        # Resume the sandbox with the tool's return value
        snapshot: ZapcodeSnapshot = state["snapshot"]
        state = snapshot.resume(result)

    return state["output"]


def run_agent(user_query: str):
    """Ask Claude to write code, then execute it in Zapcode."""
    client = anthropic.Anthropic()

    response = client.messages.create(
        model="claude-sonnet-4-20250514",
        max_tokens=1024,
        system=SYSTEM_PROMPT,
        messages=[{"role": "user", "content": user_query}],
    )

    code = response.content[0].text
    print(f"Claude wrote:\n{code}\n")

    result = execute_in_sandbox(code, {"userQuery": user_query})
    print(f"\nFinal answer: {result}")
    return result


def main():
    print("=== Example 1: Weather query ===\n")
    run_agent("What's the weather like in Tokyo?")

    print("\n=== Example 2: Multi-tool query ===\n")
    run_agent(
        "Compare the weather in London and New York, "
        "and find me cheap flights from London to New York for 2026-04-01."
    )


if __name__ == "__main__":
    main()
